File: backend/src/nodes.py
import os
from langchain_groq import ChatGroq
from dotenv import load_dotenv
from src.tools import get_mcp_tools
from src.state import State
from langchain_core.messages import SystemMessage, HumanMessage
from src.prompts import *

load_dotenv('.env.local')


class HR_Node:
    def __init__(self, llm):
        self.llm = llm 
 
    

    async def execute(self, state: State) -> State:
        user_query = state["messages"][-1].content
        
        # Get actual tools (async function that uses cached tools)
        tools = await get_mcp_tools()
        llm_with_tools = self.llm.bind_tools(tools)
        
        # Build messages with system prompt and conversation history
        messages = [
            SystemMessage(content=PROMPT),
            HumanMessage(content=user_query),
            *state["messages"]
        ]
          
        # Use ainvoke for async execution (required when graph uses ainvoke)
        response = await llm_with_tools.ainvoke(messages)
        
        return {"messages": [response], "user_query": user_query}

File: backend_js/src/nodes.ts
/**
 * HR Node implementation
 * Maps from Python nodes.py
 */

import { ChatGroq } from "@langchain/groq";
import { SystemMessage, HumanMessage } from "@langchain/core/messages";
import type { RunnableConfig } from "@langchain/core/runnables";
import { PROMPT } from "./prompts.js";
import { getMcpTools } from "./tools.js";
import type { State } from "./state.js";

/**
 * HR Node class
 * Maps from Python HR_Node class
 */
export class HRNode {
  private llm: ChatGroq;

  constructor(llm: ChatGroq) {
    this.llm = llm;
  }

  /**
   * Execute the HR node
   * Maps from Python execute method
   */
  async execute(state: State, config?: RunnableConfig): Promise<Partial<State>> {
    const lastMessage = state.messages[state.messages.length - 1];
    const userQuery = typeof lastMessage?.content === "string" 
      ? lastMessage.content 
      : typeof lastMessage?.content === "object" && Array.isArray(lastMessage.content)
      ? lastMessage.content.map(c => typeof c === "string" ? c : c.text || "").join("")
      : "";
    const jobTitle = state.job_title || "";
    const employeeId = state.employee_id || "";

    // Get actual tools (async function that uses cached tools)
    const tools = await getMcpTools();
    const llmWithTools = this.llm.bindTools(tools);

    // Include job title and employee ID in user query context for authorization decisions
    let enhancedQuery = userQuery;
    if (jobTitle) {
      enhancedQuery = `[User Job Title: ${jobTitle}, Employee ID: ${employeeId}]\n\n${userQuery}`;
    }

    // Build messages with system prompt and conversation history
    const messages = [
      new SystemMessage({ content: PROMPT }),
      new HumanMessage({ content: enhancedQuery }),
      ...state.messages,
    ];

    // Use invoke for async execution
    const response = await llmWithTools.invoke(messages, config);

    return {
      messages: [response],
      user_query: userQuery,
      job_title: jobTitle,
    };
  }
}

